plugins/generic/recommendBySimilarity/RecommendBySimilarityPlugin.inc.php

```diff
@@ -56,45 +62,87 @@ function getDescription() {
 	//
 	/**
 	 * @see templates/article/footer.tpl
+	 * @todo Revert back to using the ArticleSearch::retrieveResults() once its performance is addressed
 	 */
-	function callbackTemplateArticlePageFooter($hookName, $params) {
-		$smarty =& $params[1];
+	function callbackTemplateArticlePageFooter(string $hookName, $params) {
+		/** @var Smarty */
+		$smarty = $params[1];
 		$output =& $params[2];
 
 		// Identify similarity terms for the given article.
 		$displayedArticle = $smarty->getTemplateVars('article');
 		$articleId = $displayedArticle->getId();
 		import('classes.search.ArticleSearch');
+		import('classes.search.ArticleSearchIndex');
 		$articleSearch = new ArticleSearch();
-		$searchTerms = $articleSearch->getSimilarityTerms($articleId);
-		if (empty($searchTerms)) return false;
+		$keywords = (new ArticleSearchIndex())->filterKeywords($articleSearch->getSimilarityTerms($articleId));
+		$keywords = array_filter(array_unique($keywords), 'strlen');
+		if (!count($keywords)) {
+			return false;
+		}
 
-		// If we got similarity terms then execute a search with...
-		// ... request, journal and error messages, ...
 		$request = Application::get()->getRequest();
 		$router = $request->getRouter();
 		$journal = $router->getContext($request);
-		$error = null;
-		// ... search keywords ...
-		$query = implode(' ', $searchTerms);
-		$keywords = array(null => $query);
-		// ... and pagination.
 		$rangeInfo = Handler::getRangeInfo($request, 'articlesBySimilarity');
-		$rangeInfo->setCount(RECOMMEND_BY_SIMILARITY_PLUGIN_COUNT);
-		$smarty->assign(array(
-			'articlesBySimilarity' => $articleSearch->retrieveResults(
-					$request,
-					$journal,
-					$keywords,
-					$error,
-					null, null,
-					$rangeInfo,
-					array($articleId)
-			),
-			'articlesBySimilarityQuery' => $query,
-		));
+		$rangeInfo->setCount(static::RECOMMEND_BY_SIMILARITY_PLUGIN_COUNT);
+
+		/** @var Builder */
+		$queryBuilder = (new SubmissionQueryBuilder())
```

Comment:
This probably contains some necessary evils for 3.3.0, but for future releases, it would be more maintainable if it didn't mix idioms so much (using the SubmissionQueryBuilder, extending it locally using the query builder, and extending it locally with string-based construction). Sometimes it's worse to try to re-use code when the impedance match isn't perfect, than it is  to just fully code the query in the plugin where it can be understood without having to paw through the interactions.

If the additions to `main` to get `SubmissionQueryBuilder` are non-invasive (i.e. just adding a few functions/clauses) and possibly useful elsewhere, I'd be open to a PR there as well. Another approach might be to use `SubmissionQueryBuilder` as is, fetch or chunk its results, and perform the next operations separately (if that's not prohibitive to performance, which might be a problem here).

Long story short, I'm OK to go ahead with this, particular for `stable-3_3_0` where we're unlikely to refactor things -- but this would be a real tangle for someone to have to sort out if they changed an underlying implementation that broke this.

Reply:
Okay! This was just an initial validation, as the performance is ok and you accepted to skip the `ArticleSearch::retrieveResults()`, I'll try to improve the code 👌